src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBase.cs

```diff
@@ -140,55 +140,119 @@ private protected void ValidateInvokeTarget(object? target)
             }
         }
 
-        private protected Span<object?> CheckArguments(ref StackAllocedArguments stackArgs, ReadOnlySpan<object?> parameters, Binder? binder,
-            BindingFlags invokeAttr, CultureInfo? culture, RuntimeType[] sigTypes)
+        private protected unsafe void CheckArguments(
+            Span<object?> copyOfParameters,
+            IntPtr* byrefParameters,
+            Span<bool> shouldCopyBack,
+            ReadOnlySpan<object?> parameters,
+            RuntimeType[] sigTypes,
+            Binder? binder,
+            CultureInfo? culture,
+            BindingFlags invokeAttr
+        )
         {
-            Debug.Assert(Unsafe.SizeOf<StackAllocedArguments>() == StackAllocedArguments.MaxStackAllocArgCount * Unsafe.SizeOf<object>(),
-                "MaxStackAllocArgCount not properly defined.");
             Debug.Assert(!parameters.IsEmpty);
 
-            // We need to perform type safety validation against the incoming arguments, but we also need
-            // to be resilient against the possibility that some other thread (or even the binder itself!)
-            // may mutate the array after we've validated the arguments but before we've properly invoked
-            // the method. The solution is to copy the arguments to a different, not-user-visible buffer
-            // as we validate them. n.b. This disallows use of ArrayPool, as ArrayPool-rented arrays are
-            // considered user-visible to threads which may still be holding on to returned instances.
-
-            Span<object?> copyOfParameters = (parameters.Length <= StackAllocedArguments.MaxStackAllocArgCount)
-                    ? MemoryMarshal.CreateSpan(ref stackArgs._arg0, parameters.Length)
-                    : new Span<object?>(new object?[parameters.Length]);
-
-            ParameterInfo[]? p = null;
+            ParameterInfo[]? paramInfos = null;
             for (int i = 0; i < parameters.Length; i++)
             {
+                bool copyBackArg = false;
                 object? arg = parameters[i];
-                RuntimeType argRT = sigTypes[i];
-
                 if (arg == Type.Missing)
                 {
-                    p ??= GetParametersNoCopy();
-                    if (p[i].DefaultValue == System.DBNull.Value)
+                    paramInfos ??= GetParametersNoCopy();
+                    if (paramInfos[i].DefaultValue == DBNull.Value)
+                    {
                         throw new ArgumentException(SR.Arg_VarMissNull, nameof(parameters));
-                    arg = p[i].DefaultValue!;
+                    }
+
+                    copyBackArg = true;
+                    arg = paramInfos[i].DefaultValue;
+                }
+
+                bool isValueType;
+                RuntimeType sigType = sigTypes[i];
+
+                if (ReferenceEquals(arg?.GetType(), sigType))
```

Comment:
It is done via the `?.` operator; perhaps more efficient as separate check?

Reply:
> It is done via the ?. operator; perhaps more efficient as separate check?

As in, if `arg` is `null` the call to `ReferenceEquals()` isn't called? That is very unintuitive and I'd be surprised if that works that way. My understanding of the code here is that `ReferenceEquals()` will always be called, but if `arg` is `null` then the call will reduce to: `ReferenceEquals(null, sigType)` rather than `ReferenceEquals(arg.GetType(), sigType)`. However, my point is I don't see a reason to call `ReferenceEquals()` if `arg` is `null`. Please let me know if calling `ReferenceEquals()` with a `null` has some reason.

Reply:
That makes sense to avoid that call. I also expanded that a bit and avoid calling `CheckValue()` when `null` is passed for a reference type (since there's no validation or defaulting required).